Post gui challenges building settings

argyleink opened the pull request:
Fixes #4864 

ToDo:
- [x] editorial reviews on PR
- [x] proper YouTube link (in editing)
- [ ] release on 3/17/2021

kaycebasques commented:
Please fill out `go/chrome-devrel-content-proposal` as per usual if you haven't already

robdodson commented:
@argyleink do you need the thumb.png that's part of the PR?

argyleink commented:
> @argyleink do you need the thumb.png that's part of the PR?

last post i made the thumbnails werent ready to serve from the cdn, and i really like having a thumb that looks good in the list view. is the cdn ready? i can upload it and use the thumb from there if so.

argyleink commented:
this needs to launch on wednesday (2 days).. can it get an editor review soon please?

rachelandrew commented:
@argyleink I can take this - it's ready to look over now?

argyleink commented:
yay! yep!

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
````diff
@@ -0,0 +1,912 @@
+})
+```
+
+I think this makes for a nice visual upgrade. The slider works great without
+Javascript, the `--track-fill` prop is not required. If it's important to the
+design though, server side components could set the property on the html,
+allowing Javascript to hydrate the value while also observing any user changes. 
````
```suggestion
I think this makes for a nice visual upgrade. The slider works great without
Javascript, the `--track-fill` prop is not required. 
```

I think this needs an explanation of what the JS is doing, why you needed it.

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
````diff
@@ -0,0 +1,912 @@
+}
+```
+
+The majority of these styles are around making a nice tangible looking circle
+element. Again you see the fixed background gradient there that unifies the
+dynamic colors of the thumbs, tracks and associated SVG elements.
+
````
```suggestion
The majority of these styles are to make a nice circle. 
Again you see the fixed background gradient there that unifies the
dynamic colors of the thumbs, tracks and associated SVG elements.
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
```diff
@@ -0,0 +1,912 @@
+dynamic colors of the thumbs, tracks and associated SVG elements.
+
+I separated the styles for the interaction to help isolate the `box-shadow`
+technique being user for the hover highlight:
```
```suggestion
technique being used for the hover highlight:
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
````diff
@@ -0,0 +1,912 @@
+}
+```
+
+The goal was a nice painted highlight on the user interactive element. By using
+a box shadow that's not blurred and matches the circular shape of the thumb
+element, we can change and transition it's spread size, resulting in an easy to
+manage visual highlight effect. 
````
```suggestion
The goal was an east to manage painted highlight on the user interactive element. By using
a box shadow that's not blurred and matches the circular shape of the thumb
element, we can change and transition it's spread size. 
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
```diff
@@ -0,0 +1,912 @@
+
+## Custom Checkbox
+
+Given the following HTML input element, I'll show you how I customized it's
```
```suggestion
Given the following HTML input element, I'll show you how I customized its
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
```diff
@@ -0,0 +1,912 @@
+
+It's important to provide labels for checkboxes for 2 reasons. The first is to
+represent what the checkbox value is used for, to answer "on or off for what?".
+Second is for UX, web user's have become accustomed to interacting with
```
```suggestion
Second is for UX, web users have become accustomed to interacting with
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
```diff
@@ -0,0 +1,912 @@
+
+### HTML Attributes
+
+The icon next to the slider in my form, that's not useful to a screenreader,
+let's help it hide that from the flow. The label is also not needed for
+screenreaders, let's remove it from that flow as well:
```
```suggestion
We can hide elements that are not needed by screenreaders, 
in this case the icon next to the slider and the label:
```

feels like maybe this needs a reference as to why they are not needed.

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
```diff
@@ -0,0 +1,912 @@
+What's happening? When `max-width` is specified, it's providing context,
+explicit sizing or [definite
+sizing](https://drafts.csswg.org/css-sizing-3/#definite) for the [`auto-fit`
+layout algorythym](https://drafts.csswg.org/css-grid/#auto-repeat) to know how
```
```suggestion
layout algorithm](https://drafts.csswg.org/css-grid/#auto-repeat) to know how
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
```diff
@@ -0,0 +1,912 @@
+
+1. [Layouts](#layouts)
+1. [Color](#color)
+1. [Custom Range Input](#custom-range)
+1. [Custom Checkbox Input](#custom-checkbox)
+1. [Accessibility Considerations](#accessibility)
```
```suggestion
1. [Custom range input](#custom-range)
1. [Custom checkbox input](#custom-checkbox)
1. [Accessibility considerations](#accessibility)
```

rachelandrew reviewed (changes requested):
I've done a first pass at this. It would be a lot easier to initially review in a Google doc if it's possible in future, especially for something this long. The GH editing interface is more designed for code comments!

My main comment would be that there is a lot of stuff mentioned here that isn't supported in browsers, and there is no mention of that. It's something called out in the handbook https://web.dev/handbook/content-checklist/#compatibility

It would also be great to see new things linked to MDN at first mention. Quite a bit of this goes beyond basic webdev knowledge in as much as it is new stuff.

It's all first person, I'm kind of assuming that's ok as this is based on the video (and I found a previous one in the series https://web.dev/building-a-stories-component/). The guidelines generally ask for third person which is why I'm pointing it out.

Once I've made a few suggestions it becomes impossible to look at the thing as a whole to spot word issues. If you could resolve this round I'll jump on it again before it is published just to check for stuff I missed.

argyleink commented:
> My main comment would be that there is a lot of stuff mentioned here that isn't supported in browsers, and there is no mention of that. It's something called out in the handbook https://web.dev/handbook/content-checklist/#compatibility

I've added mentions! I'll add more too, let me know where I missed them.

> It would also be great to see new things linked to MDN at first mention. Quite a bit of this goes beyond basic webdev knowledge in as much as it is new stuff.

I'll add MDN links all day! I added a few, def note where else you'd like to see them. 

> Once I've made a few suggestions it becomes impossible to look at the thing as a whole to spot word issues. If you could resolve this round I'll jump on it again before it is published just to check for stuff I missed.

ready for another go 👍🏻

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
````diff
@@ -0,0 +1,959 @@
+}
+```
+
+Besides the pseudo element `transform-style` and `position` styles, it's mostly
+minor opinionated style stuff from me. I like the cursor to be pointer, I like
+outline offsets, default checkboxes are too tiny, and if `accent-color` is
+[supported](https://drafts.csswg.org/css-ui-4/#widget-accent), bring these
+checkboxes into the brand color scheme.
````
```suggestion
The `transform-style` and `position` styles prepare for the pseudo-element we will introduce later 
to style the highlight. Otherwise, it's mostly
minor opinionated style stuff from me. I like the cursor to be pointer, I like
outline offsets, default checkboxes are too tiny, and if `accent-color` is
[supported](https://drafts.csswg.org/css-ui-4/#widget-accent), bring these
checkboxes into the brand color scheme.
```

Is this edit correct? I couldn't figure out what you meant about the pseudo-element, but I think this is the ::before later on.

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
```diff
@@ -0,0 +1,959 @@
+### Checkbox labels
+
+It's important to provide labels for checkboxes for 2 reasons. The first is to
+represent what the checkbox value is used for, to answer "on or off for what?".
```
```suggestion
represent what the checkbox value is used for, to answer "on or off for what?"
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
```diff
@@ -0,0 +1,959 @@
+It's important to provide labels for checkboxes for 2 reasons. The first is to
+represent what the checkbox value is used for, to answer "on or off for what?".
+Second is for UX, web users have become accustomed to interacting with
+checkbox's via their associated labels.
```
```suggestion
checkboxes via their associated labels.
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
```diff
@@ -0,0 +1,959 @@
+  muted="true" 
+%}
+
+It's definitely a microinteraction, but important to me to keep the visual
+consistency. The animation scaling technique is the same as we've been using in
+other places, we set a custom property to a new value and let CSS transition it
+based on motion preferences, but it's special feature is `translateZ(-1px)`. The
+parent created a 3D space and this speudo element child tapped into it by
+placing itself slightly back in z-space.
```
```suggestion
It's definitely a micro interaction, but important to me to keep the visual
consistency. The animation scaling technique is the same as we've been using in
other places. We set a custom property to a new value and let CSS transition it
based on motion preferences. The key feature here is `translateZ(-1px)`. The
parent created a 3D space and this pseudo-element child tapped into it by
placing itself slightly back in z-space.
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
```diff
@@ -0,0 +1,959 @@
+  width="480px" 
+%}
+
+The above video demonstrates the screenreader flow on Mac OS. Notice how input
+focus is straight from one slider to the next, this is because we've hidden the
+elements that may have otherwise been stops along the way to the next element.
+Without these attributes a user would need to stop, listen and move past the
+picture and duplicative label element.
```
```suggestion
The above video demonstrates the screenreader flow on Mac OS. Notice how input
focus moves straight from one slider to the next. This is because we've hidden the
elements that may have become stops on the way to the next element.
Without these attributes a user would need to stop, listen and move past the
picture and duplicate label element.
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
````diff
@@ -0,0 +1,959 @@
+}
+```
+
+The main element is using `place-content: center` [alignment
+shorthand](https://developer.mozilla.org/en-US/docs/Web/CSS/place-content) so
+that regardless of a 2 or 1 column layout, the children should be centered
+vertically and horizontally, together as a group.
````
```suggestion
The main element is using `place-content: center` [alignment
shorthand](https://developer.mozilla.org/en-US/docs/Web/CSS/place-content) so
that the children are centered vertically and horizontally in both one and two column layouts.
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
```diff
@@ -0,0 +1,959 @@
+
+#### Centering wrapping content
+
+A common learning moment with CSS layout is distinguishing centering of `items`
+vs `content`. Flexbox and grid both provide abilities to `align-items` or
+`align-content`, and when dealing with wrapping elements, `content` layout
+alignments will distribute space amongst the children as a group. 
```
```suggestion
Flexbox and grid both provide abilities to `align-items` or
`align-content`, and when dealing with wrapping elements, `content` layout
alignments will distribute space amongst the children as a group. 
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
```diff
@@ -0,0 +1,959 @@
+  width="480px"
+%}
+
+1. Highlight context
+1. UI feedback of "how full" the value is in the range
+1. UI feedback that it's accepting input
```
```suggestion
There are many layers of UI feedback and interaction in the above video, which
help give personality to the interaction by:
- Highlighting context.
- Providing UI feedback of "how full" the value is in the range.
- Providing UI feedback that a field is accepting input.
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
```diff
@@ -0,0 +1,959 @@
+1. UI feedback of "how full" the value is in the range
+1. UI feedback that it's accepting input
+
+There are many layers of UI feedback and interaction in the above video, which
+help give personality to the interaction. Here, CSS is using the
+[`:focus-within`](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-within)
+pseudo class to change the appearance of various elements, let's break down the
+`.fieldset-item`, it's super interesting:
```
```suggestion
To provide feedback when an element is being interacted with, CSS is using the
[`:focus-within`](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-within)
pseudo class to change the appearance of various elements, let's break down the
`.fieldset-item`, it's super interesting:
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
````diff
@@ -0,0 +1,959 @@
+}
+```
+
+When this element has focus is within of one of it's children:
+1. `.fieldset-item` background is assigned a higher contrast surface color
+1. The nested `svg` is filled white for higher contrast
+1. The nested `<picture>` `clip-path` expands to a full circle and the
+   background is filled with the bright fixed gradient
````
```suggestion
When one of the children of this element has focus-within:
1. The `.fieldset-item` background is assigned a higher contrast surface color.
1. The nested `svg` is filled white for higher contrast.
1. The nested `<picture>` `clip-path` expands to a full circle and the
   background is filled with the bright fixed gradient.
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
```diff
@@ -0,0 +1,959 @@
+[PostCSS Preset Env](https://preset-env.cssdb.org/features). 
+{% endAside %}
+
+The goal was an easy to manage and animated visual highlight for user feedback.
+By using a box shadow I can prevent effecting layout with the effect. I do this
```
```suggestion
The goal was an easy to manage and animated visual highlight for user feedback.
By using a box shadow I can avoid triggering layout with the effect. I do this
```

rachelandrew replied:
might be good to add a link about what this means, maybe https://web.dev/animations-guide/ or there might be something you know of that is more focussed.

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
````diff
@@ -0,0 +1,959 @@
+}
+```
+
+Creating a circle psuedo element is straightforward work, but **placing it
+behind the element it's attached to**, that's what took work. Here's before and
+after I fixed it:
````
```suggestion
Creating a circle psuedo-element is straightforward work, but **placing it
behind the element it's attached to** was harder. Here's before and
after I fixed it:
```

rachelandrew commented on src/site/content/en/blog/building-a-settings-component/index.md:
```diff
@@ -0,0 +1,959 @@
+
+### Checkbox highlight
+
+I want to keep my interfaces consistent, and the slider element has a nice
+thumbnail highlight that I'd like to use with the checkbox. The thumbnail was
+able to use `box-shadow` and it's `spread` property to scale a shadow up and
+down, but that effect doesn't work here because our checkboxes are, [and should
+be!](https://twitter.com/argyleink/status/1329230409784291328?s=20), square.
```
```suggestion
I want to keep my interfaces consistent, and the slider element has a nice
thumbnail highlight that I'd like to use with the checkbox. The thumbnail was
able to use `box-shadow` and it's `spread` property to scale a shadow up and
down. However, that effect doesn't work here because our checkboxes are, [and should
be](https://twitter.com/argyleink/status/1329230409784291328?s=20), square.
```

rachelandrew reviewed (changes requested):
Bunch of changes to meet the styleguide, also I'm having issues reproducing your width bug as shown in the video. So I'm puzzled as to whether I'm misunderstanding what you are showing or something else. Whatever it is would be good to clarify so it is clear for anyone else!

robdodson commented:
> last post i made the thumbnails werent ready to serve from the cdn, and i really like having a thumb that looks good in the list view. is the cdn ready? i can upload it and use the thumb from there if so.

I don't think the thumb is being used in the yaml frontmatter which is why I raised it. I think you'd need to do `thumbnail: ./thumb.png`.

Taking a glance at the code for the card it looks like `thumbnail` with CDN path should be supported. Can you try uploading it and providing a similar path as you would for the `hero` frontmatter property?

argyleink commented:
> confirmed uploader and thumbnails work 👍🏻

![Screen Shot 2021-03-16 at 9 39 45 AM](https://user-images.githubusercontent.com/1134620/111346545-8eafa400-863b-11eb-82c2-05f53b19f687.png)

argyleink commented:
Rachel and I chatted and we were able to reproduce the issue together. just so there's a history note here with resolution on that editorial comment 👍🏻

mihajlija commented:
@robdodson On to the next issue: CLA for @rachelandrew. Rachel had two email address associated with her GH account and the commits on this PR are attributed to her old address, for which she can't sign the CLA. From Rachel: 

> that email address is also linked to a different Google account so I can't just add it as an alternate.

Do you know why this happens and how we can fix it?

robdodson commented:
It seems like we should remove the bit about using `aria-hidden` on the `label` element. It seems like it's maybe a quirk that it works in Chrome and VoiceOver but likely doesn't work across other browser/screen reader combos.
https://twitter.com/rob_dodson/status/1371859386210029568

argyleink commented:
> It seems like we should remove the bit about using `aria-hidden` on the `label` element

Removed it and put the learning inside a gotcha, with a link to the thread. thoughts?

rachelandrew commented:
> > It seems like we should remove the bit about using `aria-hidden` on the `label` element
> 
> Removed it and put the learning inside a gotcha, with a link to the thread. thoughts?

The prose above still refers to the label element. Should that now be phrased to just mention the picture?

argyleink commented:
> The prose above still refers to the label element. Should that now be phrased to just mention the picture?

🙂  yes, that should be more clear and have no references to the label. Updated copy 👍🏻  thoughts?

robdodson commented:
seems good now

rachelandrew reviewed (approved):
lgtm :)
